keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -0,0 +1,732 @@
+                * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+                * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly, or the volume was already mounted with a different context,
+                  volume plugin reports an error and kubelet fails to start the pod.
+                  It is CSI driver fault that it advertises SELinux support and then fails to apply it.
+
+* Nothing changes when `CSIDriver.Spec.SELinuxMountSupported` is `false` or not set:
+    * CSI volume plugin calls CSI without any special SELinux mount options and it autodetects, if the volume supports SELinux or not by presence of `seclabel` mount option.
+      This is current kubelet behavior.
+
+* Nothing changes if kubelet does not know the SELinux context of a pod (`SELinuxOptions.Level` is empty) or pod's `SELinuxRelabelPolicy` is `Always`.
+    * Volume is mounted without any SELinux options and passed to the container runtime with or without ":Z", depending on if the volume plugin supports SELinux or not (by checking `seclabel` mount option).
+      The container runtime allocates a new SELinux context and recursively relabels all files on the volume.
+      This is current kubelet behavior.
+
+Validation:
+
+* Kubernetes checks that `SELinuxRelabelPolicy` field can be used in a pod only when at least `SELinuxOptions.Level` is set.
+
+When a Pod specifies incomplete SELinux label (i.e. omits `SELinuxOptions.User`, `.Role` or `.Type`), kubelet fills the blanks from the system defaults provided by [ContainerLabels() from go-selinux bindings](https://github.com/opencontainers/selinux/blob/621ca21a5218df44259bf5d7a6ee70d720a661d5/go-selinux/selinux_linux.go#L770).
+
+### Shared volumes
+
+If a single PV that supports SELinux labels is being shared by multiple pods, each of them must have the same SELinux context.
+Currently, a running pod with context `A` will lose access to all files on a volume if a pod with context `B` starts and uses the same volume, because the container runtime relabels the volume for pod `B`.
+This behavior changes with this KEP: kubelet mounts the volume with `-o context=A` for  the first pod.
+It tries to do the same for the second pod with `-o context=B`, however, the volume has already been mounted and `mount -o context=B` fails.
+Pod `B` can't start on the same node until pod `A` dies and kubelet unmounts its volumes.
+
+We don't think that this is a bug in the design.
+Only one pod will have access to the volume, this KEP only changes the selection.
+
+The only regression is when two pods with different SELinux context use the same volume, but different SubPath - they were working before, as the container runtime relabeled only the subpaths, now the whole volume must have the same context.
```

Comment:
nit: This isn't a "regression". Users using the default mode (today's behavior) will continue to work. It's a difference in behavior for the new mode.

Reply:
reworded to:

> The only different behavior is when two pods with different SELinux context use the same volume, but different SubPath - they are working with `Always` policy, as the container runtime relabeled only the subpaths, with `OnVolumeMount` the whole volume must have the same context.